experimental/ragged_inference/seq_kv_cache.py

```diff
@@ -0,0 +1,170 @@
+        values=torch.full([n_ctx, d_model], value, **bf16_cuda()),
+    )
+
+
+def extend_kv_caches(
+    seq_kv_cache: List[SingleSeqKVCache],
+    active_keys: RaggedActivations,
+    active_values: RaggedActivations,
+) -> List[SingleSeqKVCache]:
+    assert seq_kv_cache[0].is_cuda
+
+    updated_seq_kv_cache = []
+    for cache, keys, values in zip(
+        seq_kv_cache, active_keys.iter_full_tensors(), active_values.iter_full_tensors()
+    ):
+
+        # Dim 1 is the context
+        new_cache = SingleSeqKVCache(
+            keys=torch.cat([cache.keys, keys], dim=0),
+            values=torch.cat([cache.values, values], dim=0),
+        )
+        updated_seq_kv_cache.append(new_cache)
+
+    return updated_seq_kv_cache
+
+
+def garbage_pad_seq_kv_cache(
+    seq_kv_cache: List[SingleSeqKVCache],
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    assert seq_kv_cache[0].is_cuda
+    dtype = seq_kv_cache[0].dtype
+    n_ctx_per_kv_cache = [seq.n_ctx for seq in seq_kv_cache]
+
+    # Create a view so that the output is (n_seqs, n_ctx_max, d_model)
+    # This should not incur an extra memcopy
+    n_seqs = len(n_ctx_per_kv_cache)
+    n_ctx_max = max(n_ctx_per_kv_cache)
+
+    padded_keys = torch.empty(
+        n_seqs,
+        n_ctx_max,
+        seq_kv_cache[0].d_model_per_gpu,
+        dtype=dtype,
+        device="cuda",
+    )
+
+    padded_values = torch.empty(
+        n_seqs,
+        n_ctx_max,
+        seq_kv_cache[0].d_model_per_gpu,
+        dtype=dtype,
+        device="cuda",
+    )
+
+    for seq_idx, seq in enumerate(seq_kv_cache):
+        padded_keys[seq_idx, : seq.n_ctx, :] = seq.keys
+        padded_values[seq_idx, : seq.n_ctx, :] = seq.values
+    return (padded_keys, padded_values)
+
+
+def garbage_pad_keys(
+    seq_kv_cache: List[SingleSeqKVCache],
+) -> torch.Tensor:
+    assert seq_kv_cache[0].is_cuda
+    dtype = seq_kv_cache[0].dtype
+    n_ctx_per_kv_cache = [seq.n_ctx for seq in seq_kv_cache]
+
+    # Create a view so that the output is (n_seqs, n_ctx_max, d_model)
+    # This should not incur an extra memcopy
+    n_seqs = len(n_ctx_per_kv_cache)
+    n_ctx_max = max(n_ctx_per_kv_cache)
+
+    padded_keys = torch.empty(
+        n_seqs,
+        n_ctx_max,
+        seq_kv_cache[0].d_model_per_gpu,
+        dtype=dtype,
+        device="cuda",
+    )
+
+    for seq_idx, seq in enumerate(seq_kv_cache):
+        padded_keys[seq_idx, : seq.n_ctx, :] = seq.keys
+    return padded_keys
+
+
+@lru_cache(maxsize=1)  # Memoize because we repeat this for consecutive resblocks
+def _create_indices(n_ctx_per_kv_cache):
+    """
+    We cache this because it requires some substantial CPU work and it's done multiple
+    times sequentially (once per resblock)
+    """
+    indices_list = []
+    ragged_idx = 0
+    max_n_ctx = max(n_ctx_per_kv_cache)
+    for n_ctx in n_ctx_per_kv_cache:
+        for idx_into_seq in range(max_n_ctx):
+            if idx_into_seq < n_ctx:
+                indices_list.append(ragged_idx)
```

Comment:
Wondering why we need O(n) append calls here...

Reply:
please submit a PR :D yes this is not optimal indeed